File: goblinframework-api/src/main/java/org/goblinframework/api/core/GoblinFutureResult.java
package org.goblinframework.api.core;

import org.jetbrains.annotations.Nullable;

public class GoblinFutureResult<T> {

  @Nullable public T result;
  @Nullable public Throwable cause;

  GoblinFutureResult(@Nullable T result, @Nullable Throwable cause) {
    this.result = result;
    this.cause = cause;
  }
}

File: goblinframework-core/src/main/java/org/goblinframework/core/concurrent/GoblinFutureResult.kt
package org.goblinframework.core.concurrent

import org.apache.commons.lang3.tuple.Pair

internal class GoblinFutureResult<T>
internal constructor(private val ret: T?, private val cause: Throwable?)
  : Pair<T, Throwable>() {

  override fun getLeft(): T? {
    return ret
  }

  override fun getRight(): Throwable? {
    return cause
  }

  override fun setValue(newValue: Throwable): Throwable {
    throw UnsupportedOperationException()
  }
}
